### PR TITLE
update credit skip to work again, update credits

### LINKED ISSUE #1236: Credit skip doesn't work
When you click, you don't skip a name but see the same again.

### PR DESCRIPTION
**Pull request by Flow86**
fixes #1236

@Flamefire  perhaps you can change it to use the animation system? I'm honest I don't have the time to look into it to find out how that works

### CODE REVIEW

File: libs/s25main/desktops/dskCredits.cpp [hunk 1]
```diff
@@ -174,16 +290,31 @@ void dskCredits::Draw_()
...
+    timer.restart();
+}
+
+void dskCredits::GotoPrevPage()
+{
+    // Prev page
+    if(this->itCurEntry == entries.begin())
+        this->itCurEntry = std::prev(entries.end());
+    else
+        --this->itCurEntry;
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.cpp:308]: To mirror the other function:

```suggestion
    if(this->itCurEntry == entries.begin())
        this->itCurEntry = entries.end();
    --this->itCurEntry;
```

BTW: Can you remove the trivial comments? `GotoPrevPage()` --> `// Prev page` is just noise

File: libs/s25main/desktops/dskCredits.cpp [hunk 2]
```diff
@@ -174,16 +290,31 @@ void dskCredits::Draw_()
...
+    timer.restart();
+}
+
 void dskCredits::DrawCredit()
 {
-    unsigned time = VIDEODRIVER.GetTickCount() - startTime;
+    const unsigned time = static_cast<unsigned>(std::chrono::duration_cast<std::chrono::milliseconds>(timer.getElapsed()).count());
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.cpp:314]: Just define `PAGE_TIME` as  a chrono type and don't convert at all

File: libs/s25main/desktops/dskCredits.cpp [hunk 3]
```diff
@@ -322,7 +453,20 @@ glArchivItem_Bitmap* dskCredits::GetCreditsImgOrDefault(const std::string& name)
...
+    return true;
+}
+
+bool dskCredits::Msg_LeftUp(const MouseCoords& /*mc*/)
+{
+    GotoNextPage();
+    return true;
+}
+
+bool dskCredits::Msg_RightUp(const MouseCoords& mc)
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.cpp:466]:
```suggestion
bool dskCredits::Msg_RightUp(const MouseCoords&)
```
or the warning will trigger

File: libs/s25main/desktops/dskCredits.h [hunk 4]
```diff
@@ -57,10 +64,10 @@ class dskCredits : public Desktop
         std::string lastLine;
         glArchivItem_Bitmap* pic;
         std::vector<Line> lines;
-        explicit CreditsEntry(std::string title, std::string lastLine = "")
+        explicit CreditsEntry(std::string title, std::string lastLine = std::string())
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.h:67]: Why this change? This is just more noise. Use `{}` if you don't like an empty string

File: libs/s25main/desktops/dskCredits.cpp [hunk 5]
```diff
@@ -82,69 +89,178 @@ dskCredits::dskCredits() : Desktop(LOADER.GetImageN("setup013", 0))
...
-    entry.lines.push_back(CreditsEntry::Line("Marius Loewe", 1));
-    entry.lines.push_back(CreditsEntry::Line("Eric Lutter", 1));
-    entry.lines.push_back(CreditsEntry::Line("Bob Kromonos Achten", 1));
-    entry.lines.push_back(CreditsEntry::Line("morlock", 1));
-    entry.lines.push_back(CreditsEntry::Line("Hans Gabathuler", 1));
-    entry.lines.push_back(CreditsEntry::Line("Jan Montag", 1));
-    entry.lines.push_back(CreditsEntry::Line("Patrick Schefczyk", 1));
-    entries.push_back(entry);
+
+    const size_t page_size = 15;
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.cpp:243]: You use 2 columns but 15 names? Why not 16 so it is 8 per column?

Also: Maybe alternate the names (1 left, 1 right, 1 left, ...) so the last page is filled top to bottom. But doesn't really matter

File: libs/s25main/desktops/dskCredits.cpp [hunk 6]
```diff
@@ -82,69 +89,178 @@ dskCredits::dskCredits() : Desktop(LOADER.GetImageN("setup013", 0))
...
+      // V                              // 
+      // W                              // 
+      "Niels Wiederanders",             // 
+      "Philipp Wohlleben"               // 
+      // X                              // 
+      // Y                              // 
+      // Z                              // 
+    };
+
+    // alphabetical again, two rows
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.cpp:240]:
```suggestion
    // alphabetical again, two columns
```

File: libs/s25main/desktops/dskCredits.h (hunk 4 above)
Comment (Flow86) [reply libs/s25main/desktops/dskCredits.h:67]: you shall not use `""` - it will/can create a copy from the constant `""`. but `{}` is propably the  best idea. will change that

File: libs/s25main/desktops/dskCredits.cpp (hunk 5 above)
Comment (Flow86) [reply libs/s25main/desktops/dskCredits.cpp:243]: because its not filled that way.
we have on the left 2 additional lines with the red "bottom" text - and below that the bobs will start walking around.
this way we have a quite good layout with the names and the red text.

File: libs/s25main/desktops/dskCredits.h (hunk 4 above)
Comment (Flamefire) [reply libs/s25main/desktops/dskCredits.h:67]: So what is the difference? And don't see any problem.

Comment (Flamefire) [reply libs/s25main/desktops/dskCredits.h:67]: At any optimization level (hence above `-O0`) the code is literally the same: https://godbolt.org/z/7dGAza

But readability is different:
- `void foo(std::string s = "")` - "Default value is an empty string"
- `void foo(std::string s = std::string())` - "Default value is string...with no value"
- `void foo(std::string s = {})` - "Default value is an default constructed...string"

So the first is easiest to parse visually. the second needs you to go through a lot of chars to see what the actual value is and the extra parens don't help and the last needs you to go back to see what thing you are creating and then see it's an empty string.

File: libs/s25main/desktops/dskCredits.cpp [hunk 7]
```diff
@@ -335,6 +479,7 @@ void dskCredits::SetActive(bool active)
     Desktop::SetActive(active);
     if(active)
     {
-        startTime = bobTime = bobSpawnTime = VIDEODRIVER.GetTickCount();
+        bobTime = bobSpawnTime = VIDEODRIVER.GetTickCount();
```
Comment (Flamefire) [inline libs/s25main/desktops/dskCredits.cpp:482]: This needs to change too as it is the current source of the bug which @Spikeone mistakenly thought of a "broken skip": There was no skipping it (re)activated the desktop on clicking it which restarts the timers.

It should only happen on the first call, or (now easier): When the timer isn't already running (as we never stop it it works)

**Comment by Flamefire** [comment]
@Flow86 Shall I finish this PR so we can get a release out or will you have time to do that soonish?

**Comment by Flow86** [comment]
I can do that soonish. Give me some time ;-)

File: libs/s25main/desktops/dskCredits.cpp (hunk 2 above)
Comment (Flow86) [reply libs/s25main/desktops/dskCredits.cpp:327]: doesnt work, you cant do an `a>b` comparison or similar with `std::chrono::milliseconds`?

File: libs/s25main/desktops/dskCredits.cpp (hunk 7 above)
Comment (Flow86) [reply libs/s25main/desktops/dskCredits.cpp:509]: doesnt happen for me now anymore?

Comment (Flamefire) [reply libs/s25main/desktops/dskCredits.cpp:510]: It still should as it is still wrong: The SetActive may happen multiple times caused by clicking. The `timer.restart` below will cause the `time` in `DrawCredits` to reset and the bobTime/bobSpawnTime reset above will cause an additional delay in spawning bobs and a lack in their animation (which likely won't be noticed though)
I'd suggest to use a timer for all 3, and here check if the timer has been started and only if not call start

Comment (Flow86) [reply libs/s25main/desktops/dskCredits.cpp:510]: the only thing which will happen is that you stop spawning new bobs if you click. Thats even kind of an easter egg ;-)
Credits

Comment (Flamefire) [reply libs/s25main/desktops/dskCredits.cpp:510]: It also has this skip-like behavior @Spikeone mentioned which is rather odd. However it looks like it is actually skipping now, is it? I'm surprised by that as from the code I think it just restarts the current credit. If it does skip, it is fine

Comment (Flow86) [reply libs/s25main/desktops/dskCredits.cpp:510]: it steps to the next credit on click now (and resets the time a new bob will be spawned, so its working
